server/implementation/src/main/java/io/smallrye/graphql/execution/event/EventEmitter.java

```diff
@@ -0,0 +1,196 @@
+     * custom elements.
+     * 
+     * @param builder as it stands
+     * @return builder modified by listener
+     */
+    public static GraphQLSchema.Builder fireBeforeSchemaBuild(GraphQLSchema.Builder builder) {
+        Optional<EventEmitter> emitter = emitter();
+        if (emitter.isPresent()) {
+            return emitter.get().emmitBeforeSchemaBuild(builder);
+        }
+        return builder;
+    }
+
+    /**
+     * This gets fired during the bootstrap phase before a new operation
+     * is being created. This allows listeners to modify the operation
+     * 
+     * @param operation as it stands
+     * @return operation possibly modified
+     */
+    public static Operation fireCreateOperation(Operation operation) {
+        Optional<EventEmitter> emitter = emitter();
+        if (emitter.isPresent()) {
+            return emitter.get().emmitCreateOperation(operation);
+        }
+        return operation;
+    }
+
+    // While execuring requests
+
+    public static void fireBeforeExecute() {
+        emitter().ifPresent((t) -> t.emmitBeforeExecute());
+    }
+
+    public static void fireOnExecuteError(String executionId, Throwable t) {
+        Optional<EventEmitter> emitter = emitter();
+        if (emitter.isPresent()) {
+            emitter.get().emmitOnExecuteError(executionId, t);
+        }
+    }
+
+    public static void fireAfterExecute() {
+        emitter().ifPresent((t) -> t.emmitAfterExecute());
+    }
+
+    public static void fireBeforeDataFetch() {
+        emitter().ifPresent((t) -> t.emmitBeforeDataFetch());
+    }
+
+    public static void fireBeforeMethodInvoke(InvokeInfo invokeInfo) throws Exception {
+        Optional<EventEmitter> emitter = emitter();
+        if (emitter.isPresent()) {
+            emitter.get().emmitBeforeDataFetchMethodInvoke(invokeInfo);
+        }
+    }
+
+    public static void fireOnDataFetchError(String executionId, Throwable t) {
+        Optional<EventEmitter> emitter = emitter();
+        if (emitter.isPresent()) {
+            emitter.get().emmitOnDataFetchError(executionId, t);
+        }
+    }
+
+    public static void fireAfterDataFetch() {
+        emitter().ifPresent((t) -> t.emmitAfterDataFetch());
+    }
+
+    private static Optional<EventEmitter> emitter() {
+        EventEmitter eventEmitter = current.get();
+        if (eventEmitter != null) {
+            return Optional.of(eventEmitter);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    private static final ServiceLoader<EventingService> eventingServices = ServiceLoader.load(EventingService.class);
+
+    private final List<EventingService> enabledServices = new ArrayList<>();
+
+    private EventEmitter(Config config) {
+        Iterator<EventingService> it = eventingServices.iterator();
+        while (it.hasNext()) {
+            EventingService eventingService = it.next();
```

Comment:
This is bad, because when you call `next()` it creates an instance of the service, so this will force the creation of all services that are listed in `META-INF/services/io.smallrye.graphql.spi.EventingService`, even if that service is disabled (for example, it creates an instance of `TracingService` when `smallrye.graphql.tracing.enabled=false`, which causes the whole deployment to break if tracing library is not on the classpath).

We could wrap this in a try-catch block and ignore services that can't be constructed - that is quite stupid, but I don't see a better way right now, because you can't call `getConfigKey()` without building an instance of the service :disappointed: